modules/build/src/main/scala/scala/build/options/BuildOptions.scala

```diff
@@ -284,15 +284,17 @@ final case class BuildOptions(
       sv match {
         case Some(sv0) =>
           val prefix           = if (sv0.endsWith(".")) sv0 else sv0 + "."
-          val matchingVersions = allVersions.filter(_.startsWith(prefix))
+          val matchingVersions = allVersions.filter(_.startsWith(prefix)).map(Version(_))
           if (matchingVersions.isEmpty)
             Left(new InvalidBinaryScalaVersionError(sv0))
           else {
             val validMaxVersions = maxSupportedScalaVersions
               .filter(_.repr.startsWith(prefix))
-            val validMatchingVersions = matchingVersions
-              .map(Version(_))
-              .filter(v => validMaxVersions.exists(v <= _))
+            val validMatchingVersions =
+              if (prefix.startsWith("3.") && prefix.length > 2) matchingVersions
+              else
+                matchingVersions
+                  .filter(v => validMaxVersions.exists(v <= _))
```

Comment:
I would have done that slightly differently here: if some valid versions match the user input, only consider those. If no valid versions match, then look at all available Scala versions. Something like
```suggestion
            val validMatchingVersions = {
              val filtered = matchingVersions.filter(v => validMaxVersions.exists(v <= _))
              if (filtered.isEmpty) matchingVersions
              else filtered
            }
```

Reply:
That would correspond to a "best effort" to use the valid versions, while still allowing non-valid ones when these can't work.

Reply:
Good point, your idea is much more universal.